### PR TITLE
[28] Basic GridView UI

### PR DESCRIPTION
**Pull request by mattThousand**
- Implements #28 (basic `GridView` UI based on concept comps in `Design` directory)
- Determines `tileDiameter` dynamically, as a function of the screen's bounds (on wider screens, tiles should be bigger, vice versa for narrower screens)
- Draws aesthetically nice circles inside each `TileView` using `CricleView`'s layer's border
- Adds a few tests to the suite (the codebase remains 100% TDD... which is making access control basically impossible, but resulting in perhaps more maintainable code? interested in feedback @ateliercw  @BrianSemiglia ) - w00t! 👍 )
simulator screen shot - iphone 7 plus - 2017-08-21 at 21 13 41

simulator screen shot - iphone 7 plus - 2017-08-21 at 21 13 59

### CODE REVIEW

File: ShinShikaku/Screens/UIViews/GridView.swift [hunk 1]
```diff
@@ -67,7 +50,29 @@ extension GridView: Bindable {
         // Bind selections to view layout
         Observable.just(selectionsForDisplay.asObservable())
             .bind(onNext: { winConditions in
-            // Layout view based on winConditions (i.e. player selections)
+                // Layout view based on winConditions (i.e. player selections)
+            })
+
+        dimensions.bind(onNext: { newDimensions in
+            // When dimensions are updated, remove all subviews and recreate the grid of TileViews
+            self.subviews.forEach({ $0.removeFromSuperview() })
```
Comment (ateliercw) [inline ShinShikaku/Screens/UIViews/GridView.swift]: in general `for...in` may be more efficient than using `forEach`

File: ShinShikaku/Screens/UIViews/TileView.swift [hunk 2]
```diff
@@ -13,10 +13,30 @@ final class TileView: UIView {
...
+        configureViews()
+        configureConstraints()
+    }
+
+}
+
+extension TileView: ViewConfigurable {
+
+    func configureConstraints() {
+        circleView.topAnchor.constraint(equalTo: topAnchor, constant: 0.0).isActive = true
```
Comment (ateliercw) [inline ShinShikaku/Screens/UIViews/TileView.swift]: If you're creating multiple constraints, the layout engine can be more efficient if you use `NSLayoutConstraint.activate(_ constraints: [NSLayoutConstraint])`

File: ShinShikaku/Screens/ViewType/LevelViewController.swift [hunk 3]
```diff
@@ -63,7 +64,10 @@ extension LevelViewController: ViewConfigurable {
         view.addSubview(gridView)
     }
 
-    func configureConstraints() {}
+    func configureConstraints() {
+        gridView.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
```
Comment (ateliercw) [inline ShinShikaku/Screens/ViewType/LevelViewController.swift]: If you're creating multiple constraints, the layout engine can be more efficient if you use NSLayoutConstraint.activate(_ constraints: [NSLayoutConstraint])

File: ShinShikaku/Screens/ViewType/StartMenuViewController.swift [hunk 4]
```diff
@@ -56,7 +56,12 @@ extension StartMenuViewController: ViewConfigurable {
         view.subviews.forEach({ $0.translatesAutoresizingMaskIntoConstraints = false })
     }
 
-    func configureConstraints() {}
+    func configureConstraints() {
+        resumeButton.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
```
Comment (ateliercw) [inline ShinShikaku/Screens/ViewType/StartMenuViewController.swift]: If you're creating multiple constraints, the layout engine can be more efficient if you use NSLayoutConstraint.activate(_ constraints: [NSLayoutConstraint])

File: ShinShikaku/Screens/UIViews/GridView.swift (hunk 1 above)
Comment (mattThousand) [inline ShinShikaku/Screens/UIViews/GridView.swift]: making the change, but I am curious about this - can you explain the underlying efficiency differences? 🤔

File: ShinShikaku/Screens/UIViews/TileView.swift (hunk 2 above)
Comment (mattThousand) [inline ShinShikaku/Screens/UIViews/TileView.swift]: doh! thanks

**Comment by mattThousand** [comment]
@ateliercw back to you, thanks for the review!

Couple of things: 
1. I'm genuinely curious about the `for`...`in` vs `forEach` choice. What's the underlying efficiency there?
2. I'd like to get your color literal branch merged first and then rebase this PR off of it

**Comment by ateliercw** [comment]
re for...in vs forEach: Calling out to a function has overhead, it probably gets inlined at a higher compiler optimization level, so it should be just as fast in a release build, but at the cost of (a very tiny bit of) compile time.

**Comment by mattThousand** [comment]
@ateliercw back to you. In the meantime, a philosophical question for you: to rebase or to merge? that is the question.

I've been rebasing for client work, which I suppose is more respectful of the commit history, but here I opted for a `merge` commit thinking that it's kind of nice to isolate the act of the merge in a single commit. 

Where do you stand?

**Comment by ateliercw** [comment]
On rebase VS merge, I tend to like rebasing if the changes are largely unrelated and the intermediate steps will still compile, and merging if there's any chance that rebasing will make any of the rebased commits no longer compile / pass tests because it does muck with the history of the commits.

**Pull request merged**